frontend/simple/js/state.js

```diff
@@ -167,38 +170,36 @@ const actions = {
       //       and throw an exception and write a log message.
       return console.error(`NOT EXPECTING EVENT!`, contractId, entry)
     }
-
     const type = entry.type
-    entry = Events[type].fromObject(entry, hash)
-    if (entry instanceof Events.HashableContract) {
+    entry = Events[type] ? Events[type].fromObject(entry, hash) : contracts[type].fromObject(entry, hash)
+    if (Events.HashableContract.prototype.isPrototypeOf(entry)) {
       console.log(`handleEvent for new ${type}:`, entry)
       if (entry.toObject().parentHash) {
         // TODO: use a global notification object to handle this and all other errors
         return console.error(`${type} has non-null parentHash!`, entry)
       }
       await db.addLogEntry(contractId, entry)
       commit('addContract', {contractId, type, data: entry.toVuexState()})
-    } else if (entry instanceof Events.HashableAction) {
-      const contractType = state.contracts[contractId]
+    } else if (Events.HashableAction.prototype.isPrototypeOf(entry)) {
+      const contractType = state.contracts[contractId].type
       console.log(`handleEvent for ${type} on ${contractType}:`, entry)
-      if (!Events[contractType].isActionAllowed(state[contractType], entry)) {
+      const contract = Events[contractType] || contracts[contractType]
+      if (!contract.isActionAllowed(state[contractType], entry)) {
         // TODO: implement isActionAllowed in all actions, and handle error better
         return console.error(`bad action ${type} on ${contractType} (${contractId}):`, entry)
       }
-      // Subscribe to your fellow group member's identity contracts upon joining
-      // TODO right an opposite series of operations for when someone leaves a group
-      if (entry instanceof Events.AcceptInvitation && entry.data.username !== state.loggedIn) {
-        let identity = await backend.lookup(entry.data.username)
-        await backend.subscribe(identity)
-        await dispatch('synchronize', identity)
-      }
       // TODO: verify each entry is signed by a group member
-      let hash = await db.addLogEntry(contractId, entry)
-      // If this is a duplicate the hash will be null
-      // This may occur if we get duplicate events over the network
-      if (!hash) { return }
-      commit(`${contractId}/${type}`, entry.data)
-
+      await db.addLogEntry(contractId, entry)
+      commit('setRecentHash', { contractId, hash: entry.toHash() })
+      // If this is a duplicate the hash will be null but we must process further due to users who share the same
+      // machine sharing data storage
+      commit(`${contractId}/${type}`, {data: entry.data, hash})
+      if (store.state[contractId]._async.length) {
+        for (let async of store.state[contractId]._async) {
+          dispatch(`${contractId}/${async.type}`, async.payload)
```

Comment:
We can't proceed ahead and claim we handled the event if you're not `await`ing on async events.